src/main/java/com/google/sps/servlets/DeleteClass.java

```diff
@@ -0,0 +1,178 @@
+          queue = (List<EmbeddedEntity>) classEntity.getProperty("studentQueue");
+
+          datastore.delete(txn, classKey);
+          txn.commit();
+          break;
+        } catch (ConcurrentModificationException e) {
+          if (retries == 0) {
+            throw e;
+          }
+          // Allow retry to occur
+          --retries;
+        } finally {
+          if (txn.isActive()) {
+            txn.rollback();
+          }
+        }
+      }
+
+      // Delete Visit entities
+      Filter classVisitFilter = new FilterPredicate("classKey", FilterOperator.EQUAL, classKey);
+      Query visitQuery = new Query("Visit").setFilter(classVisitFilter);
+
+      for (Entity elem : datastore.prepare(visitQuery.setKeysOnly()).asIterable()) {
+        datastore.delete(elem.getKey());
+      }
+
+      // Delete Wait entities
+      Filter classWaitFilter = new FilterPredicate("classKey", FilterOperator.EQUAL, classKey);
+      Query waitQuery = new Query("Wait").setFilter(classWaitFilter);
+
+      for (Entity elem : datastore.prepare(waitQuery.setKeysOnly()).asIterable()) {
+        datastore.delete(elem.getKey());
+      }
+
+      // Delete class key from every user
+      Filter registeredClassesFilter =
+          new FilterPredicate("registeredClasses", FilterOperator.EQUAL, classKey);
+      Query registeredClassesQuery = new Query("User").setFilter(registeredClassesFilter);
+      for (Entity entity : datastore.prepare(registeredClassesQuery).asIterable()) {
+        ArrayList<Key> classes = (ArrayList<Key>) entity.getProperty("registeredClasses");
+        classes.remove(classKey);
+        entity.setProperty("registeredClasses", classes);
+        datastore.put(entity);
+      }
+
+      Filter ownedClassesFilter =
+          new FilterPredicate("ownedClasses", FilterOperator.EQUAL, classKey);
+      Query ownedClassesQuery = new Query("User").setFilter(ownedClassesFilter);
+      for (Entity entity : datastore.prepare(ownedClassesQuery).asIterable()) {
+        ArrayList<Key> classes = (ArrayList<Key>) entity.getProperty("ownedClasses");
+        classes.remove(classKey);
+        entity.setProperty("ownedClasses", classes);
+        datastore.put(entity);
+      }
+
+      Filter taClassesFilter = new FilterPredicate("taClasses", FilterOperator.EQUAL, classKey);
+      Query taClassesQuery = new Query("User").setFilter(taClassesFilter);
+      for (Entity entity : datastore.prepare(taClassesQuery).asIterable()) {
+        ArrayList<Key> classes = (ArrayList<Key>) entity.getProperty("taClasses");
+        classes.remove(classKey);
+        entity.setProperty("taClasses", classes);
+        datastore.put(entity);
+      }
+
+      // Schedule environment deletion
+      Filter classEnvironFilter = new FilterPredicate("class", FilterOperator.EQUAL, classKey);
+      Query environQuery = new Query("Environment").setFilter(classEnvironFilter);
+
+      for (Entity elem : datastore.prepare(environQuery.setKeysOnly()).asIterable()) {
```

Comment:
```suggestion
      for (Entity elem : datastore.prepare(environQuery).asIterable()) {
```
If you only get the keys, then when you store this entity again with the status `deleting`, it will get rid of all of the other properties in the entity. This will cause the deletion task to fail because the `image` and `tag` properties no longer exist.